src/data.rs

```diff
@@ -0,0 +1,272 @@
+            &*NSString::from_str("sound"),
+        ];
+        let (main_button_label, actions, is_response): (&str, &[&str], bool) =
+            match &self.main_button {
+                Some(main_button) => match main_button {
+                    MainButton::SingleAction(main_button_label) => (main_button_label, &[], false),
+                    MainButton::DropdownActions(main_button_label, actions) => {
+                        (main_button_label, actions, false)
+                    }
+                    MainButton::Response(response) => (response, &[], true),
+                },
+                None => ("", &[], false),
+            };
+
+        let vals = vec![
+            NSString::from_str(main_button_label),
+            // TODO: Find a way to support NSArray as a NSDictionary Value rather than JUST NSString so I don't have to convert array to string and back
+            NSString::from_str(&actions.join(",")),
+            NSString::from_str(self.close_button.unwrap_or("")),
+            NSString::from_str(self.app_icon.unwrap_or("")),
+            NSString::from_str(self.content_image.unwrap_or("")),
+            NSString::from_str(self.group_id.unwrap_or_default()),
+            // TODO: Same as above, if NSDictionary could support multiple types, this could be a boolean
+            NSString::from_str(if is_response { "yes" } else { "" }),
+            NSString::from_str(&match self.delivery_date {
+                Some((delivery_date, _)) => delivery_date.to_string(),
+                _ => String::new(),
+            }),
+            // TODO: Same as above, if NSDictionary could support multiple types, this could be a boolean
+            NSString::from_str(match self.delivery_date {
+                Some((_, true)) => "yes",
+                _ => "",
+            }),
+            NSString::from_str(match self.sound {
+                Some(sound) if check_sound(sound) => sound,
+                _ => "_mute",
+            }),
+        ];
+        NSDictionary::from_keys_and_objects(keys, vals)
+    }
+}
+
+/// Response from the Notification
+#[derive(Debug)]
+pub enum NotificationResponse {
+    /// No interaction has occured
+    None,
+    /// User clicked on an action button with the given name
+    ActionButton(String),
+    /// User clicked on the close button with the given name
+    CloseButton(String),
+    /// User clicked the notification directly
+    Click,
+    /// User submitted text to the input text field
+    Reply(String),
+}
+
+impl NotificationResponse {
+    /// Create a NotificationResponse from the given Objective C NSDictionary
+    pub(crate) fn from_dictionary(dictionary: Id<NSDictionary<NSString, NSString>>) -> Self {
+        let dictionary = dictionary.deref();
+
+        let activation_type =
+            match dictionary.object_for(NSString::from_str("activationType").deref()) {
+                Some(str) => Some(str.deref().as_str().to_owned()),
```

Comment:
what is going on here?
1. this looks like it could be a `.map()`
2. what is going on here? is there no nicer way to get a `String` out of an `NSString`? do we need to? there is another `as_deref()` 4 lines later

Reply:
`.object_for()` isn't actually a for loop. It's searching for the value of the given key in the dictionary. (i.e. `dictionary["activationType"]`).

What I'm doing in this section is equivalent to:
```
// Find value in dictionary with key
match dictionary.object_for(
  // Key parameter needs to be an NSString, so convert "activationType" into an NSString
  NSString::from_str("activationType").deref()
) {
  // If it finds a value from that key, convert it into an owned String
  Some(str) => Some(str.deref().as_str().to_owned()),
  // Otherwise, don't
  None => None,
};
```

Since it isn't looping, I don't see much point to using `.map()` unless I'm missing something. There might be a better way to go about this, please let me know.

Reply:
I didn't mean it's a loop, but you can use a .map() instead of this match here
e.g. (untested, just as example)

```rust
let activation_type = dictionary
    .object_for(NSString::from_str("activationType").deref())
    .map(|str| Some(str.deref().as_str().to_owned()));
```